interface/ceed-basis.c

```diff
@@ -1386,6 +1383,198 @@ int CeedBasisApply(CeedBasis basis, CeedInt num_elem, CeedTransposeMode t_mode,
+      good_dims = ((t_mode == CEED_TRANSPOSE && (u_length >= num_points * num_q_comp || v_length >= num_nodes * num_comp)) ||
+                   (t_mode == CEED_NOTRANSPOSE && (v_length >= num_points * num_q_comp || u_length >= num_nodes * num_comp)));
+      break;
+    case CEED_EVAL_GRAD:
+    case CEED_EVAL_NONE:
+    case CEED_EVAL_WEIGHT:
+    case CEED_EVAL_DIV:
+    case CEED_EVAL_CURL:
+      // LCOV_EXCL_START
+      return CeedError(basis->ceed, CEED_ERROR_UNSUPPORTED, "Evaluation at arbitrary points not supported for %s", CeedEvalModes[eval_mode]);
+      // LCOV_EXCL_STOP
+  }
+  CeedCheck(good_dims, basis->ceed, CEED_ERROR_DIMENSION, "Input/output vectors too short for basis and evaluation mode");
+
+  // Backend method
+  if (basis->ApplyAtPoints) {
+    CeedCall(basis->ApplyAtPoints(basis, num_points, t_mode, eval_mode, x_ref, u, v));
+    return CEED_ERROR_SUCCESS;
+  }
+
+  // Default implementation
+  CeedCheck(basis->is_tensor_basis, basis->ceed, CEED_ERROR_UNSUPPORTED, "Evaluation at arbitrary points only supported for tensor product bases");
+  if (!basis->basis_chebyshev) {
+    // Build matrix mapping from quadrature point values to Chebyshev coefficients
+    CeedScalar       *tau, *C, *I, *chebyshev_coeffs_1d;
+    const CeedScalar *q_ref_1d;
+
+    // Build coefficient matrix
+    // -- Note: Clang-tidy needs this check because it does not understand the is_tensor_basis check above
+    CeedCheck(P_1d > 0 && Q_1d > 0, basis->ceed, CEED_ERROR_INCOMPATIBLE, "Basis dimensions are malformed");
+    CeedCall(CeedCalloc(Q_1d * Q_1d, &C));
+    CeedCall(CeedBasisGetQRef(basis, &q_ref_1d));
+    for (CeedInt i = 0; i < Q_1d; i++) {
+      const CeedScalar x = q_ref_1d[i];
+
+      C[i * Q_1d + 0] = 1.0;
+      C[i * Q_1d + 1] = 2 * x;
+      for (CeedInt j = 2; j < Q_1d; j++) {
+        C[i * Q_1d + j] = 2 * x * C[i * Q_1d + j - 1] - C[i * Q_1d + j - 2];
+      }
+    }
+
+    // Inverse of coefficient matrix
+    CeedCall(CeedCalloc(Q_1d * Q_1d, &chebyshev_coeffs_1d));
+    CeedCall(CeedCalloc(Q_1d * Q_1d, &I));
+    CeedCall(CeedCalloc(Q_1d, &tau));
+    // -- QR Factorization, C = Q R
+    CeedCall(CeedQRFactorization(basis->ceed, C, tau, Q_1d, Q_1d));
+    // -- chebyshev_coeffs_1d = R_inv Q^T
+    for (CeedInt i = 0; i < Q_1d; i++) I[i * Q_1d + i] = 1.0;
+    // ---- Apply R_inv, chebyshev_coeffs_1d = I R_inv
+    for (CeedInt i = 0; i < Q_1d; i++) {  // Row i
+      chebyshev_coeffs_1d[Q_1d * i] = I[Q_1d * i] / C[0];
+      for (CeedInt j = 1; j < Q_1d; j++) {  // Column j
+        chebyshev_coeffs_1d[j + Q_1d * i] = I[j + Q_1d * i];
+        for (CeedInt k = 0; k < j; k++) chebyshev_coeffs_1d[j + Q_1d * i] -= C[j + Q_1d * k] * chebyshev_coeffs_1d[k + Q_1d * i];
+        chebyshev_coeffs_1d[j + Q_1d * i] /= C[j + Q_1d * j];
+      }
+    }
+    // ---- Apply Q^T, chebyshev_coeffs_1d = R_inv Q^T
+    CeedCall(CeedHouseholderApplyQ(chebyshev_coeffs_1d, C, tau, CEED_NOTRANSPOSE, Q_1d, Q_1d, Q_1d, 1, Q_1d));
```

Comment:
Maybe we should just have a pseudoinverse utility, or at least a triangular solve utility so we aren't dealing with this here.

Reply:
Yeah, that would probably be nice to put in a separate PR. I'll put in on the ToDo list.

Reply:
see https://github.com/CEED/libCEED/issues/1250